src/app/static/src/app/components/figures/graphs/layout.ts

```diff
@@ -1,24 +1,57 @@
 import {TransformationProps, useTransformation} from "../transformedData";
-import {Dictionary} from "vue-router/types/router";
 import {deepCopy} from "../../../utils";
+import {Axis, Layout, SeriesDefinition} from "../../../generated";
 
 interface Props extends TransformationProps {
-    layout: Dictionary<any>
+    layout: Layout
 }
 
-export function useLayout(props: Props) {
+export function useLayout(props: Props, dataSeries: readonly SeriesDefinition[]) {
     const layout = deepCopy(props.layout);
-    const {evaluateFormula} = useTransformation(props);
 
+    evaluateShapes(layout, props);
+    evaluateRanges(layout, dataSeries);
+
+    return layout;
+}
+
+function evaluateShapes(layout: Layout, props: Props) {
+    const {evaluateFormula} = useTransformation(props);
     // Currently supports transformation to horizontal line only
     if (layout.shapes) {
         layout.shapes.map((shape: any) => {
             if (shape.type == "line" && shape.y_formula) {
-                const y = evaluateFormula(shape.y_formula);
+                const y = evaluateFormula(shape.y_formula) || 0;
```

Comment:
I'm assuming that omitting the shape if the formula evaluates to `undefined` is more complicated than it's worth? No problem if so... I guess the semantics are the same - a user might equally use a limit of zero to mean no limit.

Reply:
Hm, yeah, that would be nicer actually.